File: typescript/45.ts
function jump(nums: number[]): number {
    const dp: number[] = Array(nums.length).fill(0);

    for (let i = 0; i < nums.length; i++) {
        for (let j = i + 1; j <= i + nums[i] && j < nums.length; j++) {
            dp[j] = dp[j] === 0 ? dp[i] + 1 : Math.min(dp[j], dp[i] + 1);
        }
    }

    return dp[dp.length - 1];
};

test('45', () => {
    expect(jump([2, 3, 1, 1, 4])).toBe(2);
    expect(jump([2, 3, 0, 1, 4])).toBe(2);
});

File: python/45.py
import unittest
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        dp = [0] * len(nums)
        for i in range(len(nums)):
            for j in range(1, min(len(nums) - i, nums[i] + 1)):
                dp[i + j] = dp[i] + 1 if dp[i +
                                            j] == 0 else min(dp[i + j], dp[i] + 1)
        return dp[-1]


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual(solution.jump([2, 3, 1, 1, 4]), 2)
